src/java.base/unix/native/criuengine/criuengine.c

```diff
@@ -61,6 +61,75 @@ static int kickjvm(pid_t jvm, int code) {
     return 0;
 }
 
+static char *join_args(const char **args) {
+    size_t size = 0;
+    for (const char **argp = args; *argp; ++argp) {
+      size += strlen(*argp) * 4 + 2 + 1;
+    }
+    char *retval = malloc(size);
+    if (!retval) {
+        perror("malloc");
+        exit(1);
+    }
+    char *d = retval;
+    for (const char **argp = args; *argp; ++argp) {
+        if (d != retval) {
+            *d++ = ' ';
+        }
+        // https://unix.stackexchange.com/a/357932/296319
+        if (!strpbrk(*argp, " \t\n!\"#$&'()*,;<=>?[\\]^`{|}~")) {
+            d = stpcpy(d, *argp);
+            continue;
+        }
+        *d++ = '\'';
+        for (const char *s = *argp; *s; ++s) {
+          if (*s != '\'') {
+            *d++ = *s;
+          } else {
+            d = stpcpy(d, "'\\''");
+          }
+        }
+        *d++ = '\'';
+    }
+    *d = 0;
+    return retval;
+}
+
+static char *path_join(const char *path1, const char *path2) {
+    char *retval;
+    if (path2[0] == '/') {
+        retval = strdup(path2);
+        if (!retval) {
+            perror("strdup");
+            exit(1);
+        }
+        return retval;
+    }
+    if (asprintf(&retval, "%s/%s", path1, path2) == -1) {
+        perror("asprintf");
+        exit(1);
+    }
+    return retval;
+}
+
+static char *path_abs(const char *rel) {
```

Comment:
Not really as `realpath()` is resolving symlinks while this function does not. I agree the code is ugly, I originally expected it will be more nice in C++ but then I realized [std::filesystem::path](https://en.cppreference.com/w/cpp/filesystem/path) is `C++17` only.